File: apps/test-ui/src/pipelines.ts
// MIT License
//
// Copyright (c) 2026 Aparavi Software AG
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

// =============================================================================
// TEST-UI — Pipeline Definitions
// =============================================================================
//
// Factory functions that produce real pipeline objects for stress testing.
// Each call generates a unique project_id to avoid server-side collisions
// during concurrent runs.
//
// Pipeline structure:
//   { components: [...], source: '<source node id>', project_id: '<unique>' }
//
// Node wiring:
//   source (webhook/chat) -> processing nodes -> response
//   Each node's `input` array declares which lane it reads from which node.
// =============================================================================

// =============================================================================
// HELPERS
// =============================================================================

/** Generate a unique project ID with an identifying prefix. */
function uid(prefix: string): string {
	return `__test_${prefix}_${Date.now()}_${Math.random().toString(36).slice(2, 8)}__`;
}

// =============================================================================
// PIPELINE FACTORIES
// =============================================================================

/**
 * Echo pipeline: webhook -> response.
 * Simplest possible pipeline — data goes in and comes straight back out.
 * Workhorse for raw throughput and concurrent send stress tests.
 */
export function getEchoPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				name: 'Test webhook',
				config: { hideForm: true, mode: 'Source', type: 'webhook' },
			},
			{
				id: 'response_1',
				provider: 'response',
				config: { lanes: [] },
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('echo'),
	};
}

/**
 * Chain pipeline: webhook -> preprocessor_code -> response.
 * Adds a processing node so the server does real work per request.
 * Good for CPU stress and verifying single-stage processing.
 */
export function getChainPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				name: 'Test webhook',
				config: { hideForm: true, mode: 'Source', type: 'webhook' },
			},
			{
				id: 'preproc_1',
				provider: 'preprocessor_code',
				name: 'Pass-through',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'response_1',
				provider: 'response',
				config: { lanes: [] },
				input: [{ lane: 'text', from: 'preproc_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('chain'),
	};
}

/**
 * Fan-out pipeline: webhook -> 3x preprocessor_code (parallel) -> response.
 * All three processing nodes read from the same source, testing parallel
 * lane processing inside the engine.
 */
export function getFanOutPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				name: 'Test webhook',
				config: { hideForm: true, mode: 'Source', type: 'webhook' },
			},
			{
				id: 'preproc_1',
				provider: 'preprocessor_code',
				name: 'Fan-out A',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'preproc_2',
				provider: 'preprocessor_code',
				name: 'Fan-out B',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'preproc_3',
				provider: 'preprocessor_code',
				name: 'Fan-out C',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'response_1',
				provider: 'response',
				config: { lanes: [] },
				input: [
					{ lane: 'text', from: 'preproc_1' },
					{ lane: 'text', from: 'preproc_2' },
					{ lane: 'text', from: 'preproc_3' },
				],
			},
		],
		source: 'webhook_1',
		project_id: uid('fanout'),
	};
}

/**
 * Deep pipeline: webhook -> preproc -> preproc -> preproc -> preproc -> response.
 * Long chain (4 stages) that tests deep pipeline execution and event
 * propagation through many nodes.
 */
export function getDeepPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				name: 'Test webhook',
				config: { hideForm: true, mode: 'Source', type: 'webhook' },
			},
			{
				id: 'preproc_1',
				provider: 'preprocessor_code',
				name: 'Stage 1',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'preproc_2',
				provider: 'preprocessor_code',
				name: 'Stage 2',
				config: {},
				input: [{ lane: 'text', from: 'preproc_1' }],
			},
			{
				id: 'preproc_3',
				provider: 'preprocessor_code',
				name: 'Stage 3',
				config: {},
				input: [{ lane: 'text', from: 'preproc_2' }],
			},
			{
				id: 'preproc_4',
				provider: 'preprocessor_code',
				name: 'Stage 4',
				config: {},
				input: [{ lane: 'text', from: 'preproc_3' }],
			},
			{
				id: 'response_1',
				provider: 'response',
				config: { lanes: [] },
				input: [{ lane: 'text', from: 'preproc_4' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('deep'),
	};
}

/**
 * Tool pipeline: webhook -> tool_python -> response.
 * For testing client.tool() and pipe.tool() invocations against
 * the Python tool runtime.
 */
export function getToolPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				name: 'Test webhook',
				config: { hideForm: true, mode: 'Source', type: 'webhook' },
			},
			{
				id: 'tool_python_1',
				provider: 'tool_python',
				name: 'Python tool',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'response_1',
				provider: 'response',
				config: { lanes: [] },
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('tool'),
	};
}

/**
 * Chat pipeline: chat -> response.
 * Minimal chat source pipeline for client.chat() tests.
 * No LLM required — just tests the chat source path through the engine.
 */
export function getChatPipeline() {
	return {
		components: [
			{
				id: 'chat_1',
				provider: 'chat',
				name: 'Test chat',
				config: { hideForm: true, mode: 'Source', type: 'chat' },
			},
			{
				id: 'response_1',
				provider: 'response',
				config: { lanes: [] },
				input: [{ lane: 'questions', from: 'chat_1' }],
			},
		],
		source: 'chat_1',
		project_id: uid('chat'),
	};
}

// =============================================================================
// BENCHMARK PIPELINES — Real workloads from benchmark-ui
// =============================================================================

/**
 * Embedding pipeline: webhook -> preprocessor_langchain -> embedding_transformer -> response.
 * Exercises CPU (langchain chunking) and GPU (SentenceTransformer MiniLM).
 * Data: send batched text chunks, each ~100 words.
 */
export function getEmbeddingPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				config: { mode: 'Source', type: 'webhook', parameters: {} },
			},
			{
				id: 'preprocessor_1',
				provider: 'preprocessor_langchain',
				name: 'Langchain chunker',
				config: { profile: 'default' },
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'embedding_1',
				provider: 'embedding_transformer',
				name: 'Embedder',
				config: {},
				input: [{ lane: 'documents', from: 'preprocessor_1' }],
			},
			{
				id: 'response_1',
				provider: 'response_documents',
				config: { laneName: 'documents' },
				input: [{ lane: 'documents', from: 'embedding_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('embed'),
	};
}

/**
 * NER pipeline: webhook -> ner -> response.
 * GPU-heavy — GLiNER entity extraction on text.
 * Data: send individual text chunks (~150 words each).
 */
export function getNerPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				config: { mode: 'Source', type: 'webhook', parameters: {} },
			},
			{
				id: 'ner_1',
				provider: 'ner',
				name: 'Entity extraction',
				config: {},
				input: [{ lane: 'text', from: 'webhook_1' }],
			},
			{
				id: 'response_1',
				provider: 'response_text',
				config: { laneName: 'text' },
				input: [{ lane: 'text', from: 'ner_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('ner'),
	};
}

/**
 * CPU-only pipeline: webhook -> parse -> response.
 * No GPU nodes — pure CPU workload (Tika text parsing).
 * Data: send text chunks (~200 words).
 */
export function getCpuOnlyPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				config: { mode: 'Source', type: 'webhook', parameters: {} },
			},
			{
				id: 'parse_1',
				provider: 'parse',
				name: 'Tika parser',
				config: {},
				input: [{ lane: 'tags', from: 'webhook_1' }],
			},
			{
				id: 'response_1',
				provider: 'response_text',
				config: { laneName: 'text' },
				input: [{ lane: 'text', from: 'parse_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('cpuonly'),
	};
}

/**
 * Full document pipeline: webhook -> parse -> ocr -> preprocessor -> embedding -> response.
 * Most complex pipeline — exercises every processing stage.
 * Data: send PDFs or text (Tika parses files, OCR reads scanned images).
 */
export function getFullDocumentPipeline() {
	return {
		components: [
			{
				id: 'webhook_1',
				provider: 'webhook',
				config: { mode: 'Source', type: 'webhook', parameters: {} },
			},
			{
				id: 'parse_1',
				provider: 'parse',
				name: 'Tika parser',
				config: {},
				input: [{ lane: 'tags', from: 'webhook_1' }],
			},
			{
				id: 'ocr_1',
				provider: 'ocr',
				name: 'OCR',
				config: {},
				input: [{ lane: 'image', from: 'parse_1' }],
			},
			{
				id: 'preprocessor_1',
				provider: 'preprocessor_langchain',
				name: 'Langchain chunker',
				config: { profile: 'default' },
				input: [
					{ lane: 'text', from: 'parse_1' },
					{ lane: 'text', from: 'ocr_1' },
				],
			},
			{
				id: 'embedding_1',
				provider: 'embedding_transformer',
				name: 'Embedder',
				config: {},
				input: [{ lane: 'documents', from: 'preprocessor_1' }],
			},
			{
				id: 'response_1',
				provider: 'response_documents',
				config: { laneName: 'documents' },
				input: [{ lane: 'documents', from: 'embedding_1' }],
			},
		],
		source: 'webhook_1',
		project_id: uid('fulldoc'),
	};
}

// =============================================================================
// STRESS PIPELINE SELECTOR
// =============================================================================

/**
 * Stress pipeline selector — returns a different pipeline variant based on
 * the index so concurrent stress tests exercise a mix of pipeline complexities.
 *
 * Includes both simple test pipelines and real benchmark workloads.
 * Chat pipeline excluded — requires an LLM to process questions.
 *
 * @param variant - Pipeline slot index; cycled through all variants.
 */
export function getStressPipeline(variant: number) {
	const factories = [
		// Simple test pipelines (fast, no GPU needed)
		getEchoPipeline,
		getChainPipeline,
		getFanOutPipeline,
		getDeepPipeline,
		getToolPipeline,
		// Real benchmark workloads (exercise CPU + GPU processing)
		getEmbeddingPipeline,
		getNerPipeline,
		getCpuOnlyPipeline,
		getFullDocumentPipeline,
	];
	return factories[variant % factories.length]();
}

File: nodes/src/nodes/tool_pipedrive/tools/pipelines.py
# =============================================================================
# RocketRide Engine
# =============================================================================
# MIT License
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# =============================================================================

"""Pipeline and stage tools, including the conversion and movement reports."""

from __future__ import annotations

from ..pipedrive_client import clean_deal, clean_pipeline, clean_stage
from ..tool_groups import pipedrive_tool
from ._base import (
    ARR,
    BOOL,
    INT,
    NUM,
    PAGING,
    STR,
    PipedriveToolsBase,
    args_of,
    body_from,
    params_from,
    require_id,
    require_text,
    schema,
)


class PipelinesMixin(PipedriveToolsBase):
    """Tools for the ``pipelines`` and ``stages`` groups."""

    # -- pipelines --------------------------------------------------------

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(),
        description='List all pipelines. Start here to find the pipeline_id and stage layout of the account.',
    )
    def pipeline_list(self, args):
        args_of(args)
        data = self._call('GET', '/pipelines')
        return {'items': [clean_pipeline(p) for p in (data or [])]}

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(
            required=['pipeline_id'],
            pipeline_id=INT('Pipeline id.'),
            totals_convert_currency=STR('Convert deal totals to this 3-letter currency code, or "default_currency".'),
        ),
        description='Get a single pipeline, including per-stage deal totals.',
    )
    def pipeline_get(self, args):
        args = args_of(args)
        pipeline_id = require_id(args, 'pipeline_id', 'pipeline_get')
        return self._get(
            f'/pipelines/{pipeline_id}', clean_pipeline, params=params_from(args, ('totals_convert_currency',))
        )

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(
            required=['name'],
            name=STR('Pipeline name.'),
            deal_probability=BOOL('Whether deal probability is enabled for this pipeline.'),
            order_nr=INT('Display position of the pipeline.'),
            active=BOOL('Whether the pipeline is active.'),
        ),
        description='Create a pipeline.',
    )
    def pipeline_create(self, args):
        args = args_of(args)
        require_text(args, 'name', 'pipeline_create')
        return self._write(
            'POST',
            '/pipelines',
            clean_pipeline,
            body=body_from(args, ('name', 'deal_probability', 'order_nr', 'active')),
        )

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(
            required=['pipeline_id'],
            pipeline_id=INT('Pipeline id to update.'),
            name=STR('New pipeline name.'),
            deal_probability=BOOL('Whether deal probability is enabled.'),
            order_nr=INT('New display position.'),
            active=BOOL('Whether the pipeline is active.'),
        ),
        description='Update a pipeline.',
    )
    def pipeline_update(self, args):
        args = args_of(args)
        pipeline_id = require_id(args, 'pipeline_id', 'pipeline_update')
        return self._write(
            'PUT',
            f'/pipelines/{pipeline_id}',
            clean_pipeline,
            body=body_from(args, ('name', 'deal_probability', 'order_nr', 'active')),
        )

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(required=['pipeline_id'], pipeline_id=INT('Pipeline id to delete.')),
        description='Delete a pipeline.',
    )
    def pipeline_delete(self, args):
        args = args_of(args)
        return self._delete(f'/pipelines/{require_id(args, "pipeline_id", "pipeline_delete")}')

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(
            required=['pipeline_id'],
            pipeline_id=INT('Pipeline id.'),
            filter_id=INT('Apply a saved filter by id.'),
            user_id=INT('Only deals owned by this user id.'),
            stage_id=INT('Only deals in this stage.'),
            everyone=INT('Set to 1 to include deals owned by everyone, not just the authenticated user.'),
            get_summary=INT('Set to 1 to include per-stage summary totals.'),
            totals_convert_currency=STR('Convert totals to this 3-letter currency code, or "default_currency".'),
            **PAGING(),
        ),
        description='List deals in a pipeline.',
    )
    def pipeline_deals_list(self, args):
        args = args_of(args)
        pipeline_id = require_id(args, 'pipeline_id', 'pipeline_deals_list')
        return self._list(
            f'/pipelines/{pipeline_id}/deals',
            args,
            clean_deal,
            extra=params_from(
                args, ('filter_id', 'user_id', 'stage_id', 'everyone', 'get_summary', 'totals_convert_currency')
            ),
        )

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(
            required=['pipeline_id', 'start_date', 'end_date'],
            pipeline_id=INT('Pipeline id.'),
            start_date=STR('Period start, YYYY-MM-DD.'),
            end_date=STR('Period end, YYYY-MM-DD.'),
            user_id=INT('Only deals owned by this user id.'),
        ),
        description='Get stage-to-stage conversion rates and win/lost rates for a pipeline over a date range.',
    )
    def pipeline_conversion_stats(self, args):
        args = args_of(args)
        pipeline_id = require_id(args, 'pipeline_id', 'pipeline_conversion_stats')
        params = {
            'start_date': require_text(args, 'start_date', 'pipeline_conversion_stats'),
            'end_date': require_text(args, 'end_date', 'pipeline_conversion_stats'),
        }
        params.update(params_from(args, ('user_id',)))
        return self._call('GET', f'/pipelines/{pipeline_id}/conversion_statistics', params=params)

    @pipedrive_tool(
        group='pipelines',
        input_schema=schema(
            required=['pipeline_id', 'start_date', 'end_date'],
            pipeline_id=INT('Pipeline id.'),
            start_date=STR('Period start, YYYY-MM-DD.'),
            end_date=STR('Period end, YYYY-MM-DD.'),
            user_id=INT('Only deals owned by this user id.'),
        ),
        description='Get how deals moved into, through and out of each stage of a pipeline over a date range.',
    )
    def pipeline_movement_stats(self, args):
        args = args_of(args)
        pipeline_id = require_id(args, 'pipeline_id', 'pipeline_movement_stats')
        params = {
            'start_date': require_text(args, 'start_date', 'pipeline_movement_stats'),
            'end_date': require_text(args, 'end_date', 'pipeline_movement_stats'),
        }
        params.update(params_from(args, ('user_id',)))
        return self._call('GET', f'/pipelines/{pipeline_id}/movement_statistics', params=params)

    # -- stages -----------------------------------------------------------

    @pipedrive_tool(
        group='stages',
        input_schema=schema(pipeline_id=INT('Only stages of this pipeline.'), **PAGING()),
        description='List stages, optionally restricted to a single pipeline.',
    )
    def stage_list(self, args):
        args = args_of(args)
        return self._list('/stages', args, clean_stage, extra=params_from(args, ('pipeline_id',)))

    @pipedrive_tool(
        group='stages',
        input_schema=schema(
            required=['stage_id'],
            stage_id=INT('Stage id.'),
            everyone=INT('Set to 1 to include deal counts for all users.'),
        ),
        description='Get a single stage.',
    )
    def stage_get(self, args):
        args = args_of(args)
        stage_id = require_id(args, 'stage_id', 'stage_get')
        return self._get(f'/stages/{stage_id}', clean_stage, params=params_from(args, ('everyone',)))

    @pipedrive_tool(
        group='stages',
        input_schema=schema(
            required=['name', 'pipeline_id'],
            name=STR('Stage name.'),
            pipeline_id=INT('Pipeline the stage belongs to.'),
            deal_probability=NUM('Success probability percentage for deals in this stage.'),
            rotten_flag=BOOL('Whether deals in this stage can go rotten.'),
            rotten_days=INT('Days of inactivity before a deal is marked rotten.'),
            order_nr=INT('Display position of the stage in the pipeline.'),
        ),
        description='Create a stage in a pipeline.',
    )
    def stage_create(self, args):
        args = args_of(args)
        require_text(args, 'name', 'stage_create')
        require_id(args, 'pipeline_id', 'stage_create')
        body = body_from(args, ('name', 'pipeline_id', 'deal_probability', 'rotten_flag', 'rotten_days', 'order_nr'))
        return self._write('POST', '/stages', clean_stage, body=body)

    @pipedrive_tool(
        group='stages',
        input_schema=schema(
            required=['stage_id'],
            stage_id=INT('Stage id to update.'),
            name=STR('New stage name.'),
            pipeline_id=INT('Move the stage to this pipeline.'),
            deal_probability=NUM('New success probability percentage.'),
            rotten_flag=BOOL('Whether deals in this stage can go rotten.'),
            rotten_days=INT('Days of inactivity before a deal is marked rotten.'),
            order_nr=INT('New display position.'),
        ),
        description='Update a stage.',
    )
    def stage_update(self, args):
        args = args_of(args)
        stage_id = require_id(args, 'stage_id', 'stage_update')
        body = body_from(args, ('name', 'pipeline_id', 'deal_probability', 'rotten_flag', 'rotten_days', 'order_nr'))
        return self._write('PUT', f'/stages/{stage_id}', clean_stage, body=body)

    @pipedrive_tool(
        group='stages',
        input_schema=schema(required=['stage_id'], stage_id=INT('Stage id to delete.')),
        description='Delete a stage.',
    )
    def stage_delete(self, args):
        args = args_of(args)
        return self._delete(f'/stages/{require_id(args, "stage_id", "stage_delete")}')

    @pipedrive_tool(
        group='stages',
        input_schema=schema(required=['ids'], ids=ARR('Stage ids to delete.', 'integer')),
        description='Delete multiple stages in one call.',
    )
    def stage_delete_bulk(self, args):
        return self._delete_bulk('/stages', args_of(args), 'stage_delete_bulk')

    @pipedrive_tool(
        group='stages',
        input_schema=schema(
            required=['stage_id'],
            stage_id=INT('Stage id.'),
            filter_id=INT('Apply a saved filter by id.'),
            user_id=INT('Only deals owned by this user id.'),
            everyone=INT('Set to 1 to include deals owned by everyone.'),
            **PAGING(),
        ),
        description='List deals sitting in a stage.',
    )
    def stage_deals_list(self, args):
        args = args_of(args)
        stage_id = require_id(args, 'stage_id', 'stage_deals_list')
        return self._list(
            f'/stages/{stage_id}/deals', args, clean_deal, extra=params_from(args, ('filter_id', 'user_id', 'everyone'))
        )
